dbms/src/Dictionaries/ComplexKeyHashedDictionary.cpp

```diff
@@ -227,32 +227,137 @@ void ComplexKeyHashedDictionary::createAttributes()
+            for (const auto attribute_idx : ext::range(0, keys_size + attributes_size))
+            {
+                const IColumn & update_column = *block.getByPosition(attribute_idx).column.get();
+                MutableColumnPtr saved_column = saved_block->getByPosition(attribute_idx).column->mutate();
+                saved_column->insertRangeFrom(update_column, 0, update_column.size());
+            }
+        }
+
+        stream->readSuffix();
+    }
+    else
     {
-        const auto rows = block.rows();
+        auto stream = source_ptr->loadAll();
+
+        const auto saved_key_column_ptrs = ext::map<Columns>(ext::range(0, keys_size), [&](const size_t key_idx)
+        {
+            return saved_block->safeGetByPosition(key_idx).column;
+        });
+
+        stream->readPrefix();
+        while (const auto block = stream->read())
+        {
+            const auto update_key_column_ptrs = ext::map<Columns>(ext::range(0, keys_size), [&](const size_t key_idx)
+            {
+                return block.safeGetByPosition(key_idx).column;
+            });
+
+
+            std::vector<int> update_indices, saved_indices;
+            for (size_t i = 0; i < saved_block->rows(); ++i)
+            {
+                for (size_t j = 0; j < block.rows(); ++j)
+                {
+                    Arena s_temp_key_pool;
+                    Arena u_temp_key_pool;
+                    const auto s_key = placeKeysInPool(i, saved_key_column_ptrs, keys, s_temp_key_pool);
+                    const auto u_key = placeKeysInPool(j, update_key_column_ptrs, keys, s_temp_key_pool);
+                    if (s_key.toString() == u_key.toString())
+                    {
+                        update_indices.push_back(j); ///Indices of columns that are for update
+                        saved_indices.push_back(i);  ///Indices of columns that need to be updated
+                    }
+                    s_temp_key_pool.rollback(s_key.size);
+                    u_temp_key_pool.rollback(u_key.size);
+                }
+            }
+
+            auto saved_columns = saved_block->mutateColumns();
+
+            BlockPtr temp_block = std::make_shared<DB::Block>(saved_block->cloneEmpty());
+            for (const auto attribute_idx : ext::range(0, keys_size + attributes_size))
+            {
+                if (update_indices.empty())
+                    saved_columns[attribute_idx]->insertRangeFrom(*block.safeGetByPosition(attribute_idx).column, 0,
+                                                                  block.safeGetByPosition(attribute_idx).column->size());
+                else
+                {
+                    const auto &temp_columns = temp_block->mutateColumns();
+                    for (size_t i = 0; i < saved_block->rows(); ++i)
+                    {
+                        std::vector<int>::iterator it;
+                        it = std::find(saved_indices.begin(), saved_indices.end(), i);
```

Comment:
It's also O(n), and O(n*m) in total.
How to do it faster:
1. Create `IColumn::Filter` for `block` - array of '0' and '1'. Put `filter[i] = 1` if `saved_block` hasn't key from i-th row of `block`. You can build `HashTable` from `saved_block` in order to do it fast.
2. Use `IColumn::filter` for every column in `block` to remove outdated rows.
3. Insert columns from `saved_block` into `block`.

Actually, performance may be important here. For example, if dictionary contained 1M rows, it would updated probably for seconds.